moveit_ros/moveit_servo/src/servo_node.cpp

```diff
@@ -31,133 +31,220 @@
+  KinematicState next_joint_states;
 
-  if (has_error)
+  bool publish_command = false;
+
+  while (rclcpp::ok() && !stop_servo_)
   {
-    throw std::runtime_error("Servo failed to initialize : Invalid parameter values");
+    // Skip processing if servoing is disabled.
+    if (servo_paused_)
+      continue;
+
+    servo_params_ = servo_->getParams();
+    publish_command = false;
+    CommandType expectedType = servo_->expectedCommandType();
+
+    if (expectedType == CommandType::JOINT_JOG && new_joint_jog_msg_)
+    {
+      // Mark latest jointjog command as processed and also reject any twist and pose messages that had arrived simultaneously.
+      new_joint_jog_msg_ = new_twist_msg_ = new_pose_msg_ = false;
+      // JointJogCommand is an alias for VectorXd, so we can directly make a map and pass it.
+      Eigen::Map<JointJogCommand> command(latest_joint_jog_.velocities.data(), latest_joint_jog_.velocities.size());
+      next_joint_states = servo_->getNextJointState(command);
+      publish_command = (servo_->getStatus() != StatusCode::INVALID);
+    }
+    else if (expectedType == CommandType::TWIST && new_twist_msg_)
+    {
+      // Mark latest twist command as processed and also reject any jointjog and pose messages that had arrived simultaneously.
+      new_joint_jog_msg_ = new_twist_msg_ = new_pose_msg_ = false;
+      Eigen::Vector<double, 6> velocities{ latest_twist_.twist.linear.x,  latest_twist_.twist.linear.y,
+                                           latest_twist_.twist.linear.z,  latest_twist_.twist.angular.x,
+                                           latest_twist_.twist.angular.y, latest_twist_.twist.angular.z };
+      TwistCommand command{ latest_twist_.header.frame_id, velocities };
+      next_joint_states = servo_->getNextJointState(command);
+      publish_command = (servo_->getStatus() != StatusCode::INVALID);
+    }
+    else if (expectedType == CommandType::POSE && new_pose_msg_)
+    {
+      // Mark latest pose command as processed and also reject any jointjog and twist messages that had arrived simultaneously.
+      new_joint_jog_msg_ = new_twist_msg_ = new_pose_msg_ = false;
+      PoseCommand command = poseFromPoseStamped(latest_pose_);
+      next_joint_states = servo_->getNextJointState(command);
+      publish_command = (servo_->getStatus() != StatusCode::INVALID);
+    }
+    else if (new_joint_jog_msg_ || new_twist_msg_ || new_pose_msg_)
+    {
+      new_joint_jog_msg_ = new_twist_msg_ = new_pose_msg_ = false;
+      RCLCPP_WARN_STREAM(LOGGER, "Command type has not been set, cannot accept input");
+    }
+
+    if (publish_command)
+    {
+      trajectory_publisher_->publish(composeTrajectoryMessage(servo_params_, next_joint_states));
+      publish_command = false;
+    }
+    auto status_msg = std::make_unique<std_msgs::msg::Int8>();
+    status_msg->data = static_cast<int8_t>(servo_->getStatus());
+    status_publisher_->publish(std::move(status_msg));
```

Comment:
Is there a need to make this a unique pointer vs. just making the Int8 message and not having to dereference/move it to publish?

Reply:
That was copy-pasta from the old implementation.
Now that I look at the [documentation](https://docs.ros2.org/latest/api/rclcpp/classrclcpp_1_1Publisher.html), it seems creating the status once on the stack is better, since it takes a reference.

```
virtual void rclcpp::Publisher< MessageT, AllocatorT >::publish(const MessageT &  msg)
```